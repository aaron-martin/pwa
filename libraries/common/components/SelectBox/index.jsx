/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import find from 'lodash/find';
import Dropdown from '../Dropdown';
import I18n from '../I18n';
import SelectBoxItem from './components/Item';
import styles from './style';

/**
 * The select box component.
 */
class SelectBox extends Component {
  static propTypes = {
    icon: PropTypes.func.isRequired,
    item: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(PropTypes.object).isRequired,
    className: PropTypes.string,
    classNames: PropTypes.objectOf(PropTypes.string),
    defaultText: PropTypes.string,
    duration: PropTypes.number,
    handleSelectionUpdate: PropTypes.func,
    initialValue: PropTypes.string,
  };

  static defaultProps = {
    className: '',
    classNames: {},
    duration: 225,
    defaultText: 'filter.sort.default',
    handleSelectionUpdate: () => {},
    initialValue: null,
  };

  /**
   * Initializes the component.
   * @param {Object} props The components props.
   */
  constructor(props) {
    super(props);

    this.dropdownCompleted = false;
    this.state = {
      isOpen: false,
      selected: find(props.items, { value: props.initialValue }),
    };
  }

  /**
   * Reset selected when changing the initial value.
    * @param {Object} nextProps The next props the component will receive.
   */
  componentWillReceiveProps(nextProps) {
    if (this.props.initialValue !== nextProps.initialValue) {
      this.setState({
        selected: find(nextProps.items, { value: nextProps.initialValue }),
      });
    }
  }

  /**
   * When dropdown animation is finished.
   */
  onDropdownComplete = () => {
    this.dropdownCompleted = true;
  };

  /**
   * Handles any interaction the user does outside of the component.
   * In this case the select gets closed.
   * @param {Object} event The event object.
   */
  handleInteractionOutside = (event) => {
    this.setState({
      isOpen: false,
    });

    event.preventDefault();
    event.stopPropagation();
  };

  /**
   * Sets the open state to true
   */
  handleOpenList = () => {
    if (this.state.isOpen) {
      return;
    }

    this.dropdownCompleted = false;
    this.setState({
      isOpen: true,
    });
  };

  /**
   * Sets the new active selection.
   * @param {string} value Value of the selected item.
   */
  handleSelectionUpdate = (value) => {
    const selection = find(this.props.items, { value });

    this.setState({
      selected: selection,
      isOpen: false,
    });

    // Delay the callback to make sure actions are fired after animations are done.
    setTimeout(() => {
      this.props.handleSelectionUpdate(selection.value);
    }, this.props.duration);
  };

  /**
   * Renders the component
   * @returns {JSX}
   */
  render() {
    const Icon = this.props.icon;
    const { icon, selection, button, dropdown, selectItem } = this.props.classNames;

    return (
      <div className={this.props.className}>
        <button className={button} onClick={this.handleOpenList}>
          <span className={selection}>
            {this.state.selected ? (
              <I18n.Text string={this.state.selected.label} />
            ) : (
              <I18n.Text string={this.props.defaultText} />
            )}
          </span>
          {this.state.isOpen ?
            <div className={`${icon} ${styles.rotatedIcon}`}><this.props.icon /></div> :
            <div className={icon}><Icon /></div>
          }
        </button>
        <Dropdown
          className={dropdown}
          isOpen={this.state.isOpen}
          onComplete={this.onDropdownComplete}
          duration={this.props.duration}
        >
          <div>
            {this.props.items.map(item => (
              <SelectBoxItem
                className={selectItem}
                wrapper={this.props.item}
                key={item.value}
                value={item.value}
                label={item.label}
                handleSelectionUpdate={this.handleSelectionUpdate}
              />
            ))}
          </div>
        </Dropdown>
        {this.state.isOpen &&
          <button
            className={styles.overlay}
            onClick={this.handleInteractionOutside}
            onTouchMove={this.handleInteractionOutside}
          />
        }
      </div>
    );
  }
}

export default SelectBox;
