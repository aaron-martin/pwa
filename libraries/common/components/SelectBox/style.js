/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { css } from 'glamor';

const overlay = css({
  position: 'fixed',
  top: 0,
  left: 0,
  width: '100%',
  height: '100%',
  overflow: 'auto',
  zIndex: 1,
  outline: 0,
}).toString();

const rotatedIcon = css({
  transform: 'rotate(180deg)',
}).toString();

export default {
  overlay,
  rotatedIcon,
};
