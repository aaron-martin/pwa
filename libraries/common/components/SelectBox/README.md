# SelectBox
---

This is a general component used for select boxes. SelectBox component is completely
unstyled!

> **Dependents:** <br> `<FilterBar />`

## Getting Started

```
import { SelectBox } from 'Library/components';
import { ArrowDropIcon } from 'Templates/components';

<SelectBox
  items={[
    { label: 'First value', value: 'value1 },
    { label: 'Second value', value: 'value2 },
    { label: '...', value: 'value3' },
  ]}
  icon={ArrowDropIcon}
/>
```

## Styling

To style this component an object with multiple styling 'classes' can be passed.

### icon
This will be applied to the icon.

### selection
This is the styling for the span that displays the current selection.

### button
The styling for the whole button that will open/close the dropdown.

### dropdown
The wrapper of the selection list. Only visible when opened.

### selectItem
Within the dropdown each select item will get this style.

###### Usage:
```
import { css } from 'glamor';

const icon = css({ ... });
const selection = css({ ... });
const button = css({ ... });
const dropdown = css({ ... });
const selectItem = css({ ... });

const style = {
  icon,
  selection,
  button,
  dropdown,
  selectItem,
};

<SelectBox classNames={style} ... />
```

## Props

### icon (required)

_Type_: `<Icon />`

The icon will be rendered next to the current selection.

###### Usage:

```
import { ArrowDropIcon } from 'Templates/components';

<SelectBox
  items={[ ... ]}
  icon={ArrowDropIcon}
/>
```

### item (required)

_Type_: `<Component />`

A component that will be used to wrap each selection item.
This component will only receive the children prop!

###### Usage:

```
import { ArrowDropIcon, Ripple } from 'Templates/components';

const Example = (props) => (
  <Ripple fill>
    {props.children}
  </Ripple>
);

<SelectBox
  items={[ ... ]}
  item={Example}
  icon={ArrowDropIcon}
/>
```

### items (required)

_Type_: `Object[]`

`items[x].label`: `string` The label that will be displayed

`items[x].value`: `string` A unique value that identifies the selection

###### Usage

```
import { ArrowDropIcon } from 'Templates/components';

<SelectBox
  items={[
    { label: 'First value', value: 'value1 },
    { label: 'Second value', value: 'value2 },
    { label: '...', value: 'value3' },
  ]}
  icon={ArrowDropIcon}
/>
```

### classNames

_Type_: `Object`

See styling section, this can be used to style the whole component.

### defaultText

_Type_: `string`

The text that will be shown if the `initialValue` is not set.

###### Usage

```
<SelectBox
  items={[ ... ]}
  icon={ArrowDropIcon}
  defaultText="Please select..."
/>
```

### duration

_Type_: `number`
_Default_: `300`

The duration of the dropdown animation.

###### Usage

```
<SelectBox
  items={[ ... ]}
  icon={ArrowDropIcon}
  duration={500}
/>
```

### handleSelectionUpdate

_Type_: `function (newValue: number)`

Executed whenever the selection for this select box changed.
It will be called with the unique value.

###### Usage

```
<SelectBox
  items={[ ... ]}
  icon={ArrowDropIcon}
  handleSelectionUpdate={(value) => console.log(`new value: ${value}`}
/>
```

### initialValue

_Type_: `string`

When the selected value should be saved across instances this can be used
to initiate the selection. `defaultText` will be ignored when this is set.

###### Usage

```
<SelectBox
  items={[ ... ]}
  icon={ArrowDropIcon}
  initialValue="value2"
/>
```

---
