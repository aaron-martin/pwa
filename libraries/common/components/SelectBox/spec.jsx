/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { mount } from 'enzyme';
import SelectBox from './index';

/**
 * Mock Icon component.
 * @returns {JSX}
 */
const MockIconComponent = () => <span id="icon" />;

/**
 * Mock Item component.
 * @param {Object} props The components props.
 * @param {JSX} props.children The components children.
 * @returns {JSX}
 */
const MockItemComponent = ({ children }) => (
  <div>
    {children}
  </div>
);

MockItemComponent.propTypes = {
  children: PropTypes.node.isRequired,
};

describe('<SelectBox>', () => {
  const dummyItems = [
    {
      label: 'My item #1',
      value: 'item_1',
    },
    {
      label: 'My item #2',
      value: 'item_2',
    },
    {
      label: 'My item #3',
      value: 'item_3',
    },
  ];

  it('should render the selectbox with given mock components', () => {
    const wrapper = mount((
      <SelectBox icon={MockIconComponent} item={MockItemComponent} items={dummyItems} />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockIconComponent).length).toEqual(1);
    expect(wrapper.find(MockItemComponent).length).toEqual(3);
  });

  it('should render with a default text', () => {
    const wrapper = mount((
      <SelectBox
        icon={MockIconComponent}
        item={MockItemComponent}
        items={dummyItems}
        defaultText="Foo"
      />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('span').at(0).text()).toEqual('Foo');
  });

  it('should render with a preselected selection', () => {
    const wrapper = mount((
      <SelectBox
        icon={MockIconComponent}
        item={MockItemComponent}
        items={dummyItems}
        defaultText="Foo"
        initialValue="item_2"
      />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('span').at(0).text()).toEqual('My item #2');
  });
});
