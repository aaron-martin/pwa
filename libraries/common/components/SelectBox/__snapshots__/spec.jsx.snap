// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectBox> should render the selectbox with given mock components 1`] = `
<SelectBox
  className=""
  classNames={Object {}}
  defaultText="filter.sort.default"
  duration={225}
  handleSelectionUpdate={[Function]}
  icon={[Function]}
  initialValue={null}
  item={[Function]}
  items={
    Array [
      Object {
        "label": "My item #1",
        "value": "item_1",
      },
      Object {
        "label": "My item #2",
        "value": "item_2",
      },
      Object {
        "label": "My item #3",
        "value": "item_3",
      },
    ]
  }
>
  <div
    className=""
  >
    <button
      onClick={[Function]}
    >
      <span>
        <Translate
          className=""
          params={Object {}}
          string="filter.sort.default"
        >
          <span
            className=""
          >
            filter.sort.default
          </span>
        </Translate>
      </span>
      <div>
        <MockIconComponent>
          <span
            id="icon"
          />
        </MockIconComponent>
      </div>
    </button>
    <Dropdown
      className=""
      duration={225}
      easing={null}
      isOpen={false}
      onComplete={[Function]}
      onStart={[Function]}
    >
      <Transition
        duration={225}
        easing={null}
        from={
          Object {
            "height": 0,
          }
        }
        onComplete={[Function]}
        onStart={[Function]}
        origin="0 0 0"
        set={null}
        to={
          Object {
            "height": 0,
          }
        }
      >
        <div
          className="css-rdsogp "
        >
          <div>
            <SelectBoxItem
              className=""
              handleSelectionUpdate={[Function]}
              label="My item #1"
              value="item_1"
              wrapper={[Function]}
            >
              <button
                className=""
                onClick={[Function]}
              >
                <MockItemComponent>
                  <div>
                    <Translate
                      className=""
                      params={Object {}}
                      string="My item #1"
                    >
                      <span
                        className=""
                      >
                        My item #1
                      </span>
                    </Translate>
                  </div>
                </MockItemComponent>
              </button>
            </SelectBoxItem>
            <SelectBoxItem
              className=""
              handleSelectionUpdate={[Function]}
              label="My item #2"
              value="item_2"
              wrapper={[Function]}
            >
              <button
                className=""
                onClick={[Function]}
              >
                <MockItemComponent>
                  <div>
                    <Translate
                      className=""
                      params={Object {}}
                      string="My item #2"
                    >
                      <span
                        className=""
                      >
                        My item #2
                      </span>
                    </Translate>
                  </div>
                </MockItemComponent>
              </button>
            </SelectBoxItem>
            <SelectBoxItem
              className=""
              handleSelectionUpdate={[Function]}
              label="My item #3"
              value="item_3"
              wrapper={[Function]}
            >
              <button
                className=""
                onClick={[Function]}
              >
                <MockItemComponent>
                  <div>
                    <Translate
                      className=""
                      params={Object {}}
                      string="My item #3"
                    >
                      <span
                        className=""
                      >
                        My item #3
                      </span>
                    </Translate>
                  </div>
                </MockItemComponent>
              </button>
            </SelectBoxItem>
          </div>
        </div>
      </Transition>
    </Dropdown>
  </div>
</SelectBox>
`;

exports[`<SelectBox> should render with a default text 1`] = `
<SelectBox
  className=""
  classNames={Object {}}
  defaultText="Foo"
  duration={225}
  handleSelectionUpdate={[Function]}
  icon={[Function]}
  initialValue={null}
  item={[Function]}
  items={
    Array [
      Object {
        "label": "My item #1",
        "value": "item_1",
      },
      Object {
        "label": "My item #2",
        "value": "item_2",
      },
      Object {
        "label": "My item #3",
        "value": "item_3",
      },
    ]
  }
>
  <div
    className=""
  >
    <button
      onClick={[Function]}
    >
      <span>
        <Translate
          className=""
          params={Object {}}
          string="Foo"
        >
          <span
            className=""
          >
            Foo
          </span>
        </Translate>
      </span>
      <div>
        <MockIconComponent>
          <span
            id="icon"
          />
        </MockIconComponent>
      </div>
    </button>
    <Dropdown
      className=""
      duration={225}
      easing={null}
      isOpen={false}
      onComplete={[Function]}
      onStart={[Function]}
    >
      <Transition
        duration={225}
        easing={null}
        from={
          Object {
            "height": 0,
          }
        }
        onComplete={[Function]}
        onStart={[Function]}
        origin="0 0 0"
        set={null}
        to={
          Object {
            "height": 0,
          }
        }
      >
        <div
          className="css-rdsogp "
        >
          <div>
            <SelectBoxItem
              className=""
              handleSelectionUpdate={[Function]}
              label="My item #1"
              value="item_1"
              wrapper={[Function]}
            >
              <button
                className=""
                onClick={[Function]}
              >
                <MockItemComponent>
                  <div>
                    <Translate
                      className=""
                      params={Object {}}
                      string="My item #1"
                    >
                      <span
                        className=""
                      >
                        My item #1
                      </span>
                    </Translate>
                  </div>
                </MockItemComponent>
              </button>
            </SelectBoxItem>
            <SelectBoxItem
              className=""
              handleSelectionUpdate={[Function]}
              label="My item #2"
              value="item_2"
              wrapper={[Function]}
            >
              <button
                className=""
                onClick={[Function]}
              >
                <MockItemComponent>
                  <div>
                    <Translate
                      className=""
                      params={Object {}}
                      string="My item #2"
                    >
                      <span
                        className=""
                      >
                        My item #2
                      </span>
                    </Translate>
                  </div>
                </MockItemComponent>
              </button>
            </SelectBoxItem>
            <SelectBoxItem
              className=""
              handleSelectionUpdate={[Function]}
              label="My item #3"
              value="item_3"
              wrapper={[Function]}
            >
              <button
                className=""
                onClick={[Function]}
              >
                <MockItemComponent>
                  <div>
                    <Translate
                      className=""
                      params={Object {}}
                      string="My item #3"
                    >
                      <span
                        className=""
                      >
                        My item #3
                      </span>
                    </Translate>
                  </div>
                </MockItemComponent>
              </button>
            </SelectBoxItem>
          </div>
        </div>
      </Transition>
    </Dropdown>
  </div>
</SelectBox>
`;

exports[`<SelectBox> should render with a preselected selection 1`] = `
<SelectBox
  className=""
  classNames={Object {}}
  defaultText="Foo"
  duration={225}
  handleSelectionUpdate={[Function]}
  icon={[Function]}
  initialValue="item_2"
  item={[Function]}
  items={
    Array [
      Object {
        "label": "My item #1",
        "value": "item_1",
      },
      Object {
        "label": "My item #2",
        "value": "item_2",
      },
      Object {
        "label": "My item #3",
        "value": "item_3",
      },
    ]
  }
>
  <div
    className=""
  >
    <button
      onClick={[Function]}
    >
      <span>
        <Translate
          className=""
          params={Object {}}
          string="My item #2"
        >
          <span
            className=""
          >
            My item #2
          </span>
        </Translate>
      </span>
      <div>
        <MockIconComponent>
          <span
            id="icon"
          />
        </MockIconComponent>
      </div>
    </button>
    <Dropdown
      className=""
      duration={225}
      easing={null}
      isOpen={false}
      onComplete={[Function]}
      onStart={[Function]}
    >
      <Transition
        duration={225}
        easing={null}
        from={
          Object {
            "height": 0,
          }
        }
        onComplete={[Function]}
        onStart={[Function]}
        origin="0 0 0"
        set={null}
        to={
          Object {
            "height": 0,
          }
        }
      >
        <div
          className="css-rdsogp "
        >
          <div>
            <SelectBoxItem
              className=""
              handleSelectionUpdate={[Function]}
              label="My item #1"
              value="item_1"
              wrapper={[Function]}
            >
              <button
                className=""
                onClick={[Function]}
              >
                <MockItemComponent>
                  <div>
                    <Translate
                      className=""
                      params={Object {}}
                      string="My item #1"
                    >
                      <span
                        className=""
                      >
                        My item #1
                      </span>
                    </Translate>
                  </div>
                </MockItemComponent>
              </button>
            </SelectBoxItem>
            <SelectBoxItem
              className=""
              handleSelectionUpdate={[Function]}
              label="My item #2"
              value="item_2"
              wrapper={[Function]}
            >
              <button
                className=""
                onClick={[Function]}
              >
                <MockItemComponent>
                  <div>
                    <Translate
                      className=""
                      params={Object {}}
                      string="My item #2"
                    >
                      <span
                        className=""
                      >
                        My item #2
                      </span>
                    </Translate>
                  </div>
                </MockItemComponent>
              </button>
            </SelectBoxItem>
            <SelectBoxItem
              className=""
              handleSelectionUpdate={[Function]}
              label="My item #3"
              value="item_3"
              wrapper={[Function]}
            >
              <button
                className=""
                onClick={[Function]}
              >
                <MockItemComponent>
                  <div>
                    <Translate
                      className=""
                      params={Object {}}
                      string="My item #3"
                    >
                      <span
                        className=""
                      >
                        My item #3
                      </span>
                    </Translate>
                  </div>
                </MockItemComponent>
              </button>
            </SelectBoxItem>
          </div>
        </div>
      </Transition>
    </Dropdown>
  </div>
</SelectBox>
`;
